src/Plugin/Field/FieldWidget/ProductVariationOptionsWidget.php

```diff
@@ -0,0 +1,239 @@
+    $product = $form_state->get('product');
+    $variations = $this->loadEnabledVariations($product);
+
+    if (count($variations) === 0) {
+      // Nothing to purchase, tell the parent form to hide itself.
+      $form_state->set('hide_form', TRUE);
+      $element['variation'] = [
+        '#type' => 'value',
+        '#value' => 0,
+      ];
+
+      return $element;
+    }
+
+    if (count($variations) > 1) {
+      // This widget only supports one variation.
+      $form_state->set('hide_form', TRUE);
+      $element['variation'] = [
+        '#type' => 'value',
+        '#value' => 0,
+      ];
+
+      $this->loggerFactory
+        ->get('commerce_option')
+        ->error('Product variation options widget only supports one variation. Product with the id @id has @count.', [
+          '@id' => $product->id(),
+          '@count' => count($variations),
+        ]);
+
+      return $element;
+    }
+
+    /** @var \Drupal\commerce_product\Entity\ProductVariationInterface $selected_variation */
+    $selected_variation = reset($variations);
+
+    $form_state->set('selected_variation', $selected_variation->id());
+    $this->tempStore->get('commerce_option')->set('variation_id', $selected_variation->id());
+
+    $element['variation'] = [
+      '#type' => 'value',
+      '#value' => $selected_variation->id(),
+    ];
+
+    // Build the full attribute form.
+    $wrapper_id = Html::getUniqueId('commerce-product-add-to-cart-form');
+    $form += [
+      '#wrapper_id' => $wrapper_id,
+      '#prefix' => '<div id="' . $wrapper_id . '">',
+      '#suffix' => '</div>',
+    ];
+
+    $element['options'] = [
+      '#type' => 'container',
+      '#attributes' => [
+        'class' => ['option-widgets'],
+      ],
+    ];
+
+    /** @var \Drupal\commerce_order\Entity\OrderItemInterface $order_item */
+    $order_item = $items->getEntity();
+    $purchasableEntity = $order_item->getPurchasedEntity();
+    $options = $this->optionResolver->resolveOptions($purchasableEntity);
+
+    foreach ($options as $id => $option) {
+      $element['options'][$id] = [
+        '#type' => $option->getElementType(),
+        '#title' => $option->label(),
+        '#options' => $this->getOptionValueNames($purchasableEntity, $option),
+        '#default_value' => '',
+        '#limit_validation_errors' => [],
+        '#ajax' => [
+          'callback' => [get_class($this), 'ajaxRefresh'],
+          'wrapper' => $form['#wrapper_id'],
+          // Prevent a jump to the top of the page.
+          'disable-refocus' => TRUE,
+        ],
+      ];
+    }
+
+    return $element;
+  }
+
+  /**
+   * Returns the product variation.
+   *
+   * @param \Drupal\commerce_product\Entity\ProductInterface $product
+   *   The product.
+   *
+   * @return \Drupal\commerce_product\Entity\ProductVariationInterface|null
+   *   The product variation or null. This widget only supports one variation.
+   */
+  protected function getVariation(ProductInterface $product) {
```

Comment:
Are we using this function anywhere?

Reply:
Good catch. no, looks like i removed the usage at some point. this can be removed